docs/proposals/koordlet/20231208-support-netqos.md

```diff
@@ -0,0 +1,395 @@
+	Policies *ResourceQOSPolicies `json:"policies,omitempty"`
+
+	// ResourceQOS for LSR pods.
+	LSRClass *ResourceQOS `json:"lsrClass,omitempty"`
+
+	// ResourceQOS for LS pods.
+	LSClass *ResourceQOS `json:"lsClass,omitempty"`
+
+	// ResourceQOS for BE pods.
+	BEClass *ResourceQOS `json:"beClass,omitempty"`
+
+	// ResourceQOS for system pods
+	SystemClass *ResourceQOS `json:"systemClass,omitempty"`
+
+	// ResourceQOS for root cgroup.
+	CgroupRoot *ResourceQOS `json:"cgroupRoot,omitempty"`
+}
+
+type ResourceQOS struct {
+	...
+	NetworkQOS *NetworkQOSCfg `json:"networkQOS,omitempty"`
+}
+
+type NetworkQOSCfg struct {
+	Enable     *bool `json:"enable,omitempty"`
+	NetworkQOS `json:",inline"`
+}
+
+type NetworkQOS struct {
+	// IngressRequest describes the minimum network bandwidth guaranteed in the ingress direction.
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=0
+	IngressRequest *intstr.IntOrString `json:"ingressRequest,omitempty"`
+	// IngressLimit describes the maximum network bandwidth can be used in the ingress direction,
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=100
+	IngressLimit *intstr.IntOrString `json:"ingressLimit,omitempty"`
+
+	// EgressRequest describes the minimum network bandwidth guaranteed in the egress direction.
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=0
+	EgressRequest *intstr.IntOrString `json:"egressRequest,omitempty"`
+	// EgressLimit describes the maximum network bandwidth can be used in the egress direction,
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=100
+	EgressLimit *intstr.IntOrString `json:"egressLimit,omitempty"`
+}
+
+type SystemStrategy struct {
+	...
+	// TotalNetworkBandwidth indicates the overall network bandwidth, cluster manager can set this field via "slo-controller-config" configmap, 
+	// and default value just taken from /sys/class/net/${NIC_NAME}/speed, unit: Mbps
+	TotalNetworkBandwidth resource.Quantity `json:"totalNetworkBandwidth,omitempty"`
+}
+```
+
+##### pod level：
+To enable fine-grained network bandwidth control for containers in a pod, `koordinator` will sync configuration file 
+content to `/var/run/koordinator/net/pods`. and then `terway-qos` or other netqos plugin(eg: tc) will do something to 
+limit the net bandwidth that container can use. content as follows:
```

Comment:
How to use file api and netqos plugin to interact to ensure that QoS has taken effect when the Pod is started?
 How to prevent Pod from starting if QoS configuration fails?

Reply:
l0(online) priority by default

Reply:
> l0(online) priority by default, so it will be treated as online process when QoS configuration fails.

What are the considerations for this design?
Perhaps configuring QoS should be more appropriate as part of Pod startup. Just like the CPU, if the cgroup configuration fails, the Pod will not start.

Reply:
netqos plugin itself is working as an external plugin , configuration errors will be handled by default , if we hopes pod not start when configration errors, we need handle it in `runtimehook` ,  but it is not compatible with multiple netqos implementations.

Reply:
> How to prevent Pod from starting if QoS configuration fails
@lujinda 
network qos api can be implemented as built-in plugin(in koordlet) or external plugin(e.g. terway-qos).
for built-in plugin, this error is handled by `runtimehook`, pod will start failed if NRI enabled, or get failure "pod condition/event" if NRI not enabled.
for external plugin, koordlet does not guarantee the network qos takes effective once startup, yet. koordlet will only monitor the network qos status and report "pod condition/event" if anything goes wrong.